api_factory/schema/naming.py

```diff
@@ -0,0 +1,182 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import dataclasses
+import os
+import re
+from typing import Iterable, Tuple
+
+from google.api import annotations_pb2
+from google.protobuf import descriptor_pb2
+
+from api_factory import utils
+
+
+@dataclasses.dataclass(frozen=True)
+class Naming:
+    """Naming data for an API.
+
+    This class contains the naming nomenclature used for this API
+    within templates.
+
+    An instance of this object is made available to every template
+    (as ``api.naming``).
+    """
+    name: str
+    namespace: Tuple[str]
+    version: str
+    product_name: str
+    product_url: str
+
+    @classmethod
+    def build(cls,
+            *file_descriptors: Iterable[descriptor_pb2.FileDescriptorProto]
+            ) -> 'Naming':
+        """Return a full APINaming instance based on these file descriptors.
+
+        This is pieced together from the proto package names as well as the
+        ``google.api.metadata`` file annotation. This information may be
+        present in one or many files; this method is tolerant as long as
+        the data does not conflict.
+
+        Args:
+            file_descriptors (Iterable[~.FileDescriptorProto]): A list of
+                file descriptor protos. This list should only include the
+                files actually targeted for output (not their imports).
+
+        Returns:
+            ~.Naming: A :class:`~.Naming` instance which is provided to
+                templates as part of the :class:`~.API`.
+
+        Raises:
+            ValueError: If the provided file descriptors contain contradictory
+                information.
+        """
+        # Determine the set of proto packages.
+        proto_packages = {fd.package for fd in file_descriptors}
+        root_package = os.path.commonprefix(tuple(proto_packages)).rstrip('.')
+
+        # Sanity check: If there is no common ground in the package,
+        # we are obviously in trouble.
+        if not root_package:
+            raise ValueError('Protos provided have entirely different '
+                             'proto packages.')
+
+        # Define the valid regex to split the package.
+        #
+        # It is not necessary for the regex to be as particular about package
+        # name validity (e.g. avoiding .. or segments starting with numbers)
+        # because protoc is guaranteed to give us valid package names.
+        pattern = r'^((?P<namespace>[a-z0-9_.]+)\.)?(?P<name>[a-z0-9_]+)'
+
+        # Only require the version portion of the regex if the version is
+        # present.
+        #
+        # This code may look counter-intuitive (why not use ? to make it
+        # optional), but the engine's greediness routine will decide that
+        # the version is the name, which is not what we want.
```

Comment:
Note: the engine's greediness can be limited by appending `?` to `+` or `*`. If the namespace pattern is changed to `(?P<namespace>[a-z0-9_.]+?)` it will never eat more than it's supposed to.